### PR TITLE
Account page serde memory

### PR DESCRIPTION
**Pull request by arhimondr**
## Description

When compression and encryption enabled page serialization / deserialization requires additional buffering space (up to 128kB). This PR ensures this memory is properly accounted. 

## Additional context and related issues

https://github.com/trinodb/trino/issues/15273

## Release notes

(X) This is not user-visible or docs only and no release notes are required.
( ) Release notes are required, please propose a release note for me.
( ) Release notes are required, with the following suggested text:

```markdown
# Section
* Fix some things. ({issue}`issuenumber`)
```

### CODE REVIEW

File: core/trino-main/src/main/java/io/trino/operator/MergeOperator.java [hunk 1]
```diff
@@ -147,6 +148,11 @@ public MergeOperator(
         this.comparator = requireNonNull(comparator, "comparator is null");
         this.outputChannels = requireNonNull(outputChannels, "outputChannels is null");
         this.outputTypes = requireNonNull(outputTypes, "outputTypes is null");
+
+        LocalMemoryContext memoryContext = operatorContext.newLocalUserMemoryContext(MergeOperator.class.getSimpleName());
+        // memory footprint of deserializer does not change over time
+        memoryContext.setBytes(deserializer.getRetainedSizeInBytes());
+        closer.register(memoryContext::close);
```
Comment (losipiuk) [inline core/trino-main/src/main/java/io/trino/operator/MergeOperator.java:155]: why so? We call close() on `finish`. And memory is still not reclaimed yet.

Also you do not use this pattern in `ExchangeOperator` why a difference?

File: core/trino-main/src/main/java/io/trino/spiller/FileSingleStreamSpiller.java [hunk 2]
```diff
@@ -60,8 +62,8 @@
 
     private final FileHolder targetFile;
     private final Closer closer = Closer.create();
-    private final PageSerializer serializer;
-    private final PageDeserializer deserializer;
+    private final PagesSerdeFactory serdeFactory;
```
Comment (losipiuk) [inline core/trino-main/src/main/java/io/trino/spiller/FileSingleStreamSpiller.java:65]: In commit messge you wrote:

```
Given that spilling streams are not expected to be tiny there is very
little benefit from caching serializer / deserializer.
```

If they are not tiny, then the caching is a benefit, right?
Or did you mean that a single call to `writePages` is not short?

**Review by losipiuk — APPROVED** [review]

File: core/trino-main/src/main/java/io/trino/spiller/FileSingleStreamSpiller.java (hunk 2 above)
Comment (arhimondr) [reply core/trino-main/src/main/java/io/trino/spiller/FileSingleStreamSpiller.java:65]: I mean a single call to writePages / readPages is not short, but the time spend between these two calls could in theory be long. @findepi could you please take a look and let me know if my understanding is correct?

File: core/trino-main/src/main/java/io/trino/operator/MergeOperator.java (hunk 1 above)
Comment (arhimondr) [reply core/trino-main/src/main/java/io/trino/operator/MergeOperator.java:155]: Great catch.

I honestly don't know why are we calling `close` in `finish`.

`finish` indicates that no more pages will be added to operator via `addInput`, and it is meaningless for source operators (such as `ExchangeOperator`, `MergeOperator`) and it should not get called (and from what I understand it doesn't). 

Regardless, today both of these operators simply call `close` on `finish` what effectively means operator is no longer usable. 

Perhaps we should follow up with an another PR to resolve this confusion.

**Comment by arhimondr** [comment]
Another bug: https://github.com/trinodb/trino/pull/15277/commits/cb31a780489892dbf651f253450d25133796d499

In ExchangeOperator the memory is now reserved at the moment of the operator creation. When a downstream operator fails during creation already created operators don't get properly closed what results in a memory leak. 

I also considered avoiding reserving memory in a constructor, but regardless it doesn't seem right not to close operators hence I decided to keep memory allocation in a constructor.

**Comment by arhimondr** [comment]
Updated

File: core/trino-main/src/main/java/io/trino/operator/DriverFactory.java [hunk 3]
```diff
@@ -96,12 +97,37 @@ public synchronized Driver createDriver(DriverContext driverContext)
...
+                try {
+                    operator.close();
+                }
+                catch (Throwable closeFailure) {
+                    if (failure != closeFailure) {
+                        failure.addSuppressed(closeFailure);
+                    }
+                }
+            }
+            for (OperatorContext operatorContext : driverContext.getOperatorContexts()) {
```
Comment (losipiuk) [inline core/trino-main/src/main/java/io/trino/operator/DriverFactory.java:119]: Do we know that at this point the operator context which are present in driverContext only correspond to operators created above?

**Review by losipiuk — APPROVED** [review]

File: core/trino-main/src/main/java/io/trino/operator/DriverFactory.java (hunk 3 above)
Comment (arhimondr) [reply core/trino-main/src/main/java/io/trino/operator/DriverFactory.java:119]: It is possible that Operator creation fails after the corresponding `OperatorContext` has already been created

Comment (losipiuk) [reply core/trino-main/src/main/java/io/trino/operator/DriverFactory.java:119]: Yeah - that I understand. My quesiton is can there be context from other operators, which should not be closed here?

Comment (arhimondr) [reply core/trino-main/src/main/java/io/trino/operator/DriverFactory.java:119]: There shouldn't be any other OperatorContext's but the ones for a driver that we failed to created

Comment (losipiuk) [reply core/trino-main/src/main/java/io/trino/operator/DriverFactory.java:119]: Yeah - looked into the code. I though there should be 1:1 mapping between Driver and DriverContext, just it looked weird that both are not created at the same time. I guess some optimization.

**Review by losipiuk — APPROVED** [review]

**Pull request merged**